Gitweb:     http://git.kernel.org/git/?p=linux/kernel/git/torvalds/linux-2.6.git;a=commit;h=538452700d95480c16e7aa6b10ff77cd937d33f4
Commit:     538452700d95480c16e7aa6b10ff77cd937d33f4
Parent:     9e42d0cf5020aaf217433cad1a224745241d212a
Author:     NeilBrown <neilb@suse.de>
AuthorDate: Fri Jan 9 08:31:05 2009 +1100
Committer:  NeilBrown <neilb@suse.de>
CommitDate: Fri Jan 9 08:31:05 2009 +1100

    md: fix bitmap-on-external-file bug.
    
    commit a2ed9615e3222645007fc19991aedf30eed3ecfd
    fixed a bug with 'internal' bitmaps, but in the process broke
    'in a file' bitmaps.  So they are broken in 2.6.28
    
    This fixes it, and needs to go in 2.6.28-stable.
    
    Signed-off-by: NeilBrown <neilb@suse.de>
    Cc: stable@kernel.org

 drivers/md/bitmap.c |    8 +++++---
 1 files changed, 5 insertions(+), 3 deletions(-)

diff --git a/drivers/md/bitmap.c b/drivers/md/bitmap.c
index ab7c8e4..666b7ba 100644
--- a/drivers/md/bitmap.c
+++ b/drivers/md/bitmap.c
@@ -964,9 +964,11 @@ static int bitmap_init_from_disk(struct bitmap *bitmap, sector_t start)
 				 */
 				page = bitmap->sb_page;
 				offset = sizeof(bitmap_super_t);
-				read_sb_page(bitmap->mddev, bitmap->offset,
-					     page,
-					     index, count);
+				if (!file)
+					read_sb_page(bitmap->mddev,
+						     bitmap->offset,
+						     page,
+						     index, count);
 			} else if (file) {
 				page = read_page(file, index, bitmap, count);
 				offset = 0;
